### DISCUSSION TITLE
How to deploy patched/updated SWT version?

### DISCUSSION THREAD

**Discussion by phermsdorf** [eclipse-tycho/tycho#2279] [Q&A]
Hi,

we are in the need to deploy a patched version of the SWT fragment org.eclipse.swt.win32.win32.x86_64.

We currently use the SWT release belonging to Eclipse 2022-06. The updated/patched fragment has version 3.123.100.v20230322-1206 (SNAPSHOT build version from CI build because not yet released) and it has the following constraint to the Host Bundle:

Fragment-Host: org.eclipse.swt;bundle-version="[3.116.0,4.0.0)"

which would basically work because the version of our current org.eclipse.swt in our target platform is 3.120.0.v20220530-1036. From within the IDE it is possible to launch the application with that swt version.

But I'm struggling on how to get that to build and update with p2.

What I tried was adding the new Fragment to the target and filtering out the current version. That does not work for multiple reasons:

1) the org.eclipse.e4.rcp feature includes (of course) the concrete current version of org.eclipse.swt.win32.win32.x86_64. So i had to remove the filtering of the target again

2) but then the build fails with 
```
[ERROR]   Unresolved requirement: Import-Package: org.eclipse.swt.widgets
[ERROR]     -> Export-Package: org.eclipse.swt.widgets; bundle-symbolic-name="org.eclipse.swt.win32.win32.x86_64"; bundle-version="3.123.100.v20230322-1206"; version="0.0.0"
[ERROR]        org.eclipse.swt.win32.win32.x86_64 [13]
[ERROR]          Unresolved requirement: Fragment-Host: org.eclipse.swt; bundle-version="[3.116.0,4.0.0)"
[ERROR]          Unresolved requirement: Require-Capability: eclipse.platform; filter:="(& (osgi.ws=win32) (osgi.os=win32) (osgi.arch=x86_64))"
```

where 3.123.100 is the new updated version of org.eclipse.swt.win32.win32.x86_64.
I believe that may have to do with that org.eclipse.swt seems to have a p2.inf file which requires/enforces the exact same fragment version like the host bundle ...

My idea was to include the new version of the Fragment, require the installation by including this fragment into one of our bundles and _hope_ that at runtime the newer version is used (like in the IDE).
But the build fails with

`[ERROR]   Only one of the following can be installed at once: [org.eclipse.swt.win32.win32.x86_64 3.123.100.v20230322-1206, org.eclipse.swt.win32.win32.x86_64 3.120.0.v20220530-1036]
`
because these Fragment's are singleton's.

Even if i would patch class files in the current jar p2 would not trigger an update since the version stay's the same.

Any ideas or hints on how to solve such a problem?

Thanks,
Bye Peter

**Comment by laeubi** [comment]
I fear this can't be resolved easily, first you could try to not use the `org.eclipse.e4.rcp` feature but simply include whats needed in your custom feature, beside that, you should open an issue to E4 not include SWT in that feature as it will be required transitively already and to prevent such issue in the future.

Next you should edit the swt fragment (if we assume its till compatible) to have a wider version range for the host and use that in your target, then the build should work. As an alternative you can try to simply use the full updated SWT version, not only a single fragment.

**Comment by phermsdorf** [comment]
> Next you should edit the swt fragment (if we assume its till compatible) to have a wider version range for the host and use that in your target, then the build should work. As an alternative you can try to simply use the full updated SWT version, not only a single fragment.

I tried with a custom/copied feature and included org.eclipse.swt, org.eclipse.swt.win32.win32.x86_64 in the target and removed the conrete versions in the feature (version = 0.0.0).
And i added a filter like this to the target-platform-configuration
```
<filter>
    <type>eclipse-plugin</type>
    <id>org.eclipse.swt</id>
    <restrictTo>
        <versionRange>[3.123,4)</versionRange>
    </restrictTo>
</filter>
```
but the build fails with
```
 Unresolved requirement: Import-Package: org.eclipse.swt.widgets
    -> Export-Package: org.eclipse.swt.widgets; bundle-symbolic-name="org.eclipse.swt.win32.win32.x86_64"; bundle-version="3.123.100.v20230322-1206"; version="0.0.0"
       org.eclipse.swt.win32.win32.x86_64 [13]
         Unresolved requirement: Fragment-Host: org.eclipse.swt; bundle-version="[3.116.0,4.0.0)"
         Unresolved requirement: Require-Capability: eclipse.platform; filter:="(& (osgi.ws=win32) (osgi.os=win32) (osgi.arch=x86_64))"
```
but for me a boundary definition like "[3.116.0,4.0.0)" (in the fragment) should match the bundle org.eclipse.swt version "3.123.100.v20230322-1206" which exists in the target platform...

Any ideas?

**Reply by phermsdorf** [reply]
@laeubi 
does filters (https://tycho.eclipseprojects.io/doc/latest/target-platform-configuration/target-platform-configuration-mojo.html#filters) of type eclipse-plugin also apply to fragments?

**Reply by laeubi** [reply]
I think so but in general I try to avoid filters at all as they are not portable and show something is wrong in your target setup, as usually the constraints of imports should be enough to select appropriate dependencies.
So I suspect that if you need to filter this more shows that somewhere there is actually still a dependency to the older swt.

**Reply by phermsdorf** [reply]
yeah right, but i'm still not able to find it. 
the error 'Unresolved requirement: Fragment-Host: org.eclipse.swt; bundle-version="[3.116.0,4.0.0)"' still persists even that i think that i have a version 3.123.100.v20230322-1206 in place ...

**Reply by laeubi** [reply]
Even if you remove the filter? I think the best would be to try creating a very small reproducer project that shows the error in isolation.

**Reply by phermsdorf** [reply]
i tried that based on the demo/itp04-rcp project. See here
https://github.com/phermsdorf/tycho/tree/issue_use_newer_swt_version
to simulate the newer SWT version i added the repository https://download.eclipse.org/eclipse/updates/4.28-I-builds/ to the mars.target file and removed the eclipse.rcpfeatures and copied thei content basically to feature1 and changed the version consraints of org.eclipse.swt and the 2 fragments for linux and windows. 

Adding the repository seems to work in the "Generic Text Editor" even with version completion etc. but PDE as well as the tycho build stops with
```
[ERROR]   Problems resolving provisioning plan.:
[ERROR]      org.eclipse.swt.gtk.linux.x86_64 3.123.100.v20230321-0918 cannot be installed in this environment because its filter is not applicable.
```
which i don't really understand

**Reply by laeubi** [reply]
> which i don't really understand

You usually only include the swt bundle not the fragments, the suitable fragment is then selected automatically!

**Reply by phermsdorf** [reply]
Of course! What a dumb mistake. That always was included in some fragment. 
I'll try to get the example running and come back here.
Thanks for the hint!

**Comment by phermsdorf** [comment]
Ok with that change the sample project (and our's) seem to build fine.
For reference i updated the sample with my adjustments [here](https://github.com/phermsdorf/tycho/tree/issue_use_newer_swt_version)

Conclusion:
* removed eclipse.rcp and eclipse.e4.rcp feature
* created separate rcp feature with their content and removed version boundaries (set to 0.0.0) on the swt bundle and fragments

Thanks for your help!

bye Peter
